// vite-plugin-css-injected-by-js/dist/utils.js
import { build } from 'vite';
import { warnLog, debugLog } from './utils.log.js';
const cssInjectedByJsId = '\0vite/all-css';
const defaultInjectCode = (cssCode, { useStrictCSP, attributes }) => {
    let attributesInjection = '';
    for (const attribute in attributes) {
        const attributeValue = typeof attributes[attribute] === 'function' ? attributes[attribute]() : attributes[attribute];
        attributesInjection += `elementStyle.setAttribute('${attribute}', '${attributeValue}');`;
    }
    return `try{if(typeof document != 'undefined'){var elementStyle = document.createElement('style');${useStrictCSP ? `elementStyle.nonce = document.head.querySelector('meta[property=csp-nonce]')?.content;` : ''}${attributesInjection}elementStyle.appendChild(document.createTextNode(${cssCode}));document.head.appendChild(elementStyle);}}catch(e){console.error('vite-plugin-css-injected-by-js', e);}`;
};
export async function buildCSSInjectionCode({ buildOptions, cssToInject, injectCode, injectCodeFunction, injectionCodeFormat = 'iife', useStrictCSP, attributes, }) {
    let { minify, target } = buildOptions;
    const res = await build({
        root: '',
        configFile: false,
        logLevel: 'error',
        plugins: [
            injectionCSSCodePlugin({
                cssToInject,
                injectCode,
                injectCodeFunction,
                useStrictCSP,
                attributes,
            }),
        ],
        build: {
            write: false,
            target,
            minify,
            assetsDir: '',
            rollupOptions: {
                input: {
                    ['all-css']: cssInjectedByJsId,
                },
                output: {
                    format: injectionCodeFormat
                },
            },
        },
    });
    const _cssChunk = Array.isArray(res) ? res[0] : res;
    if (!('output' in _cssChunk))
        return null;
    // Rolldown emits //#region and //#endregion line-comments that reference
    // internal module IDs.  When the injection code is later collapsed to a
    // single line, these // comments swallow all subsequent code on that line,
    // producing invalid JS.  Strip them here so the output is safe to flatten.
    const code = _cssChunk.output[0].code.replace(/^\s*\/\/#(?:region|endregion).*$/gm, '');
    return { ..._cssChunk.output[0], code };
}
export function resolveInjectionCode(cssCode, injectCode, injectCodeFunction, { useStrictCSP, attributes }) {
    const injectionOptions = { useStrictCSP, attributes };
    if (injectCodeFunction) {
        return `(${injectCodeFunction})(${cssCode}, ${JSON.stringify(injectionOptions)})`;
    }
    const injectFunction = injectCode || defaultInjectCode;
    return injectFunction(cssCode, injectionOptions);
}
function injectionCSSCodePlugin({ cssToInject, injectCode, injectCodeFunction, useStrictCSP, attributes, }) {
    return {
        name: 'vite:injection-css-code-plugin',
        resolveId(id) {
            if (id == cssInjectedByJsId) {
                return id;
            }
        },
        load(id) {
            if (id == cssInjectedByJsId) {
                const cssCode = JSON.stringify(cssToInject.trim());
                return resolveInjectionCode(cssCode, injectCode, injectCodeFunction, { useStrictCSP, attributes });
            }
        },
    };
}
export function removeLinkStyleSheets(html, cssFileName) {
    const removeCSS = new RegExp(`<link rel=".*"[^>]*?href=".*/?${cssFileName}"[^>]*?>`);
    return html.replace(removeCSS, '');
}
export { warnLog, debugLog } from './utils.log.js';
function isJsOutputChunk(chunk) {
    return chunk.type == 'chunk' && chunk.fileName.match(/.[cm]?js(?:\?.+)?$/) != null;
}
function defaultJsAssetsFilter(chunk) {
    return chunk.isEntry && !chunk.fileName.includes('polyfill');
}
// The cache must be global since execution context is different every entry
const cssSourceCache = {};
export function extractCss(bundle, cssName) {
    const cssAsset = bundle[cssName];
    if (cssAsset !== undefined && cssAsset.source) {
        const cssSource = cssAsset.source;
        // We treat these as strings and coerce them implicitly to strings, explicitly handle conversion
        cssSourceCache[cssName] =
            cssSource instanceof Uint8Array ? new TextDecoder().decode(cssSource) : `${cssSource}`;
    }
    return cssSourceCache[cssName] ?? '';
}
export function concatCssAndDeleteFromBundle(bundle, cssAssets) {
    return cssAssets.reduce(function extractCssAndDeleteFromBundle(previous, cssName) {
        const cssSource = extractCss(bundle, cssName);
        delete bundle[cssName];
        return previous + cssSource;
    }, '');
}
export function buildJsCssMap(bundle, jsAssetsFilterFunction) {
    const chunksWithCss = {};
    const bundleKeys = getJsTargetBundleKeys(bundle, typeof jsAssetsFilterFunction == 'function' ? jsAssetsFilterFunction : () => true);
    if (bundleKeys.length === 0) {
        throw new Error('Unable to locate the JavaScript asset for adding the CSS injection code. It is recommended to review your configurations.');
    }
    for (const key of bundleKeys) {
        const chunk = bundle[key];
        if (chunk.type === 'asset' || !chunk.viteMetadata || chunk.viteMetadata.importedCss.size === 0) {
            continue;
        }
        const chunkStyles = chunksWithCss[key] || [];
        chunkStyles.push(...chunk.viteMetadata.importedCss.values());
        chunksWithCss[key] = chunkStyles;
    }
    return chunksWithCss;
}
export function getJsTargetBundleKeys(bundle, jsAssetsFilterFunction) {
    if (typeof jsAssetsFilterFunction != 'function') {
        const jsAssets = Object.keys(bundle).filter((i) => {
            const asset = bundle[i];
            return isJsOutputChunk(asset) && defaultJsAssetsFilter(asset);
        });
        if (jsAssets.length == 0) {
            return [];
        }
        const jsTargetFileName = jsAssets[jsAssets.length - 1];
        if (jsAssets.length > 1) {
            warnLog(`[vite-plugin-css-injected-by-js] has identified "${jsTargetFileName}" as one of the multiple output files marked as "entry" to put the CSS injection code.` +
                'However, if this is not the intended file to add the CSS injection code, you can use the "jsAssetsFilterFunction" parameter to specify the desired output file (read docs).');
            if (process.env.VITE_CSS_INJECTED_BY_JS_DEBUG) {
                const jsAssetsStr = jsAssets.join(', ');
                debugLog(`[vite-plugin-css-injected-by-js] identified js file targets: ${jsAssetsStr}. Selected "${jsTargetFileName}".\n`);
            }
        }
        // This should be always the root of the application
        return [jsTargetFileName];
    }
    const chunkFilter = ([_key, chunk]) => isJsOutputChunk(chunk) && jsAssetsFilterFunction(chunk);
    return Object.entries(bundle)
        .filter(chunkFilter)
        .map(function extractAssetKeyFromBundleEntry([key]) {
        return key;
    });
}
export async function relativeCssInjection(bundle, assetsWithCss, buildCssCode, topExecutionPriorityFlag, buildOptions, isVirtualModuleUsed = false) {
    for (const [jsAssetName, cssAssets] of Object.entries(assetsWithCss)) {
        process.env.VITE_CSS_INJECTED_BY_JS_DEBUG &&
            debugLog(`[vite-plugin-css-injected-by-js] Relative CSS: ${jsAssetName}: [ ${cssAssets.join(',')} ]`);
        const assetCss = concatCssAndDeleteFromBundle(bundle, cssAssets);
        const cssInjectionCode = assetCss.length > 0 ? (await buildCssCode(assetCss))?.code : '';
        // We have already filtered these chunks to be RenderedChunks
        const jsAsset = bundle[jsAssetName];
        injectAndFixMap(jsAsset, cssInjectionCode ?? '', assetCss, buildOptions, topExecutionPriorityFlag, bundle, isVirtualModuleUsed);
    }
}
const globalCSSCodeEntryCache = new Map();
let previousFacadeModuleId = '';
export async function globalCssInjection(bundle, cssAssets, buildCssCode, jsAssetsFilterFunction, topExecutionPriorityFlag, buildOptions, isVirtualModuleUsed = false) {
    const jsTargetBundleKeys = getJsTargetBundleKeys(bundle, jsAssetsFilterFunction);
    if (jsTargetBundleKeys.length == 0) {
        throw new Error('Unable to locate the JavaScript asset for adding the CSS injection code. It is recommended to review your configurations.');
    }
    process.env.VITE_CSS_INJECTED_BY_JS_DEBUG &&
        debugLog(`[vite-plugin-css-injected-by-js] Global CSS Assets: [${cssAssets.join(',')}]`);
    const allCssCode = concatCssAndDeleteFromBundle(bundle, cssAssets);
    let cssInjectionCode = '';
    if (allCssCode.length > 0) {
        const cssCode = (await buildCssCode(allCssCode))?.code;
        if (typeof cssCode == 'string') {
            cssInjectionCode = cssCode;
        }
    }
    for (const jsTargetKey of jsTargetBundleKeys) {
        const jsAsset = bundle[jsTargetKey];
        /**
         * Since it creates the assets once sequential builds for the same entry point
         * (for example when multiple formats of same entry point are built),
         * we need to reuse the same CSS created the first time.
         */
        if (jsAsset.facadeModuleId != null && jsAsset.isEntry && cssInjectionCode != '') {
            if (jsAsset.facadeModuleId != previousFacadeModuleId) {
                globalCSSCodeEntryCache.clear();
            }
            previousFacadeModuleId = jsAsset.facadeModuleId;
            globalCSSCodeEntryCache.set(jsAsset.facadeModuleId, cssInjectionCode);
        }
        if (cssInjectionCode == '' &&
            jsAsset.isEntry &&
            jsAsset.facadeModuleId != null &&
            typeof globalCSSCodeEntryCache.get(jsAsset.facadeModuleId) == 'string') {
            cssInjectionCode = globalCSSCodeEntryCache.get(jsAsset.facadeModuleId);
        }
        process.env.VITE_CSS_INJECTED_BY_JS_DEBUG &&
            debugLog(`[vite-plugin-css-injected-by-js] Global CSS inject: ${jsAsset.fileName}`);
        injectAndFixMap(jsAsset, cssInjectionCode, allCssCode, buildOptions, topExecutionPriorityFlag, bundle, isVirtualModuleUsed);
        /*jsAsset.code = buildOutputChunkWithCssInjectionCode(
            jsAsset.code,
            cssInjectionCode ?? '',
            topExecutionPriorityFlag
        );*/
    }
}
export function buildOutputChunkWithCssInjectionCode(jsAssetCode, cssInjectionCode, topExecutionPriorityFlag) {
    const appCode = jsAssetCode.replace(/\/\*\s*empty css\s*\*\//g, '');
    jsAssetCode = topExecutionPriorityFlag ? '' : appCode;
    jsAssetCode += cssInjectionCode;
    jsAssetCode += !topExecutionPriorityFlag ? '' : appCode;
    return jsAssetCode;
}
export function clearImportedCssViteMetadataFromBundle(bundle, unusedCssAssets) {
    // Required to exclude removed files from manifest.json
    for (const key in bundle) {
        const chunk = bundle[key];
        if (chunk.viteMetadata && chunk.viteMetadata.importedCss.size > 0) {
            const importedCssFileNames = chunk.viteMetadata.importedCss;
            importedCssFileNames.forEach((importedCssFileName) => {
                if (!unusedCssAssets.includes(importedCssFileName) && chunk.viteMetadata) {
                    chunk.viteMetadata.importedCss = new Set();
                }
            });
        }
    }
}
export function isCSSRequest(request) {
    const CSS_LANGS_RE = /\.(css|less|sass|scss|styl|stylus|pcss|postcss|sss)(?:$|\?)/;
    return CSS_LANGS_RE.test(request);
}
export function injectAndFixMap(chunk, cssInjectionCode, rawCss, // NEW PARAMETER
buildOptions, topExecutionPriority, bundle, isVirtualModuleUsed = false) {
    chunk.code = chunk.code.replace(/\/\*\s*empty css\s*\*\//g, '');
    // Check both now, since in SSR cssInjectionCode might be empty but rawCss exists
    if (!cssInjectionCode && !rawCss)
        return;
    let mapObj = null;
    const mapAssetName = chunk.fileName + '.map';
    const mapAsset = bundle[mapAssetName];
    if (buildOptions?.sourcemap && mapAsset?.type === 'asset') {
        try {
            mapObj = JSON.parse(String(mapAsset.source));
        }
        catch (_) { }
    }
    const shiftMap = () => {
        if (mapObj && mapAsset) {
            mapObj.mappings = ';' + mapObj.mappings;
            mapAsset.source = JSON.stringify(mapObj);
        }
        if (chunk.map && typeof chunk.map.mappings === 'string') {
            chunk.map.mappings = ';' + chunk.map.mappings;
        }
    };
    if (isVirtualModuleUsed) {
        const patched = cssInjectionCode ? cssInjectionCode.replace(/document\.head/g, 'document_head') : '';
        // 1. Write the payload as a REAL function for syntax highlighting and linting!
        const payloadTemplate = function () {
            /* SSR Support: Store raw CSS globally */
            if (typeof globalThis !== 'undefined') {
                globalThis.__VITE_CSS_RAW__ = (globalThis.__VITE_CSS_RAW__ || '') + '%%RAW_CSS%%';
            }
            /* DOM Injection Support */
            if (typeof document !== 'undefined' && typeof globalThis !== 'undefined') {
                var executeInject = function (options) {
                    var target = (options && options.target) || document.head;
                    if (!target)
                        return;
                    executeInject.cache = executeInject.cache || [];
                    for (var i = 0; i < executeInject.cache.length; i++) {
                        if (executeInject.cache[i].target === target) {
                            var els = executeInject.cache[i].elements;
                            for (var j = 0; j < els.length; j++)
                                target.appendChild(els[j]);
                            return;
                        }
                    }
                    var newElements = [];
                    var observer = new MutationObserver(function () { });
                    var obsTarget = target.nodeType === 11 ? target : (document.documentElement || document);
                    observer.observe(obsTarget, { childList: true, subtree: true });
                    try {
                        (function (document_head) {
                            // %%PATCHED_CODE%%
                        })(target);
                    }
                    finally {
                        var records = observer.takeRecords();
                        globalThis.__VITE_CSS_ELS__ = globalThis.__VITE_CSS_ELS__ || [];
                        for (var i = 0; i < records.length; i++) {
                            for (var j = 0; j < records[i].addedNodes.length; j++) {
                                var node = records[i].addedNodes[j];
                                newElements.push(node);
                                globalThis.__VITE_CSS_ELS__.push({ target: target, el: node });
                            }
                        }
                        observer.disconnect();
                    }
                    executeInject.cache.push({ target: target, elements: newElements });
                };
                globalThis.__VITE_CSS_QUEUE__ = globalThis.__VITE_CSS_QUEUE__ || [];
                globalThis.__VITE_CSS_QUEUE__.push(executeInject);
            }
        };
        // 2. Serialize the function to a string, and replace our placeholders!
        const payload = `(${payloadTemplate.toString()})();`
            .replace("'%%RAW_CSS%%'", JSON.stringify(rawCss || ''))
            .replace('// %%PATCHED_CODE%%', patched);
        const singleLine = payload.replace(/\n/g, '').replace(/\s{2,}/g, ' ');
        // Decision 1: ALWAYS put at the top for virtual module to ensure queue is ready
        chunk.code = singleLine + '\n' + chunk.code;
        shiftMap();
    }
    else {
        const singleLine = cssInjectionCode.replace(/\n/g, '');
        if (topExecutionPriority) {
            chunk.code = singleLine + '\n' + chunk.code;
            shiftMap();
        }
        else {
            chunk.code += '\n' + singleLine;
        }
    }
}
